### ISSUE TITLE
[BUG] Different Numerical Result When Using 'qunat', 'scale' and 'covar'

### ISSUE THREAD

**Issue by myc634** [nerfstudio-project/gsplat#630]
Hi,

Thanks for your great work!

I am trying to render gaussian with gsplat, however, I get a slightly lower numerical result when using the `quant` and `scale` compare to using `covar`. This kind of pattern happens a lot and affect my final result. 

For example, I get 

![Image](https://github.com/user-attachments/assets/6a2ec3dc-4d16-4632-9d68-cdae5f8c0e7a)

when I use `quant` and `scale` as the input of the rasterizer, however, I get

![Image](https://github.com/user-attachments/assets/820a3abb-fcc4-49fa-964c-30fe67b0088d)

when I add extra parameter `covar` as the input.

Here is the code I use to convert `quant` and `scale`  to `covar`:

```
def quaternion_to_matrix(
    quaternions: Float[Tensor, "*batch 4"],
    eps: float = 1e-8,
) -> Float[Tensor, "*batch 3 3"]:
    # Order changed to match scipy format!
    i, j, k, r = torch.unbind(quaternions, dim=-1)
    two_s = 2 / ((quaternions * quaternions).sum(dim=-1) + eps)

    o = torch.stack(
        (
            1 - two_s * (j * j + k * k),
            two_s * (i * j - k * r),
            two_s * (i * k + j * r),
            two_s * (i * j + k * r),
            1 - two_s * (i * i + k * k),
            two_s * (j * k - i * r),
            two_s * (i * k - j * r),
            two_s * (j * k + i * r),
            1 - two_s * (i * i + j * j),
        ),
        -1,
    )
    return rearrange(o, "... (i j) -> ... i j", i=3, j=3)


def build_covariance(
    scale: Float[Tensor, "*#batch 3"],
    rotation_xyzw: Float[Tensor, "*#batch 4"],
) -> Float[Tensor, "*batch 3 3"]:
    scale = scale.diag_embed()
    rotation = quaternion_to_matrix(rotation_xyzw)
    return (
        rotation
        @ scale
        @ rearrange(scale, "... i j -> ... j i")
        @ rearrange(rotation, "... i j -> ... j i")
    )
```

And the mean result on 140 samples suffers 0.5db lower on PSNR metric

**Comment by liruilong940607** [comment]
This is unexpected! Will try to fine some cycles look into this. If you would be able to provide a minimal code snip that captures the difference in the forward/backward pass with some inputs (a few GS maybe) that will be very helpful for me to start.

**Comment by myc634** [comment]
Sure, could you please provide the assets that you need for debugging? I can save them with .pt format and upload into google drive so that you can access

**Comment by myc634** [comment]
Actually, I am modifying the code of [NopoSplat](https://github.com/cvg/NoPoSplat), the following is the modification:

I replace the forward function here: https://github.com/cvg/NoPoSplat/blob/9b04307ebe179d610c04208db8d69f7c3106d03b/src/model/decoder/decoder_splatting_cuda.py#L37

with the following code:

```
def forward(
        self,
        gaussians: Gaussians,
        extrinsics: Float[Tensor, "batch view 4 4"],
        intrinsics: Float[Tensor, "batch view 3 3"],
        near: Float[Tensor, "batch view"],
        far: Float[Tensor, "batch view"],
        image_shape: tuple[int, int],
        depth_mode: DepthRenderingMode | None = None,
        cam_rot_delta: Float[Tensor, "batch view 3"] | None = None,
        cam_trans_delta: Float[Tensor, "batch view 3"] | None = None,
    ) -> DecoderOutput:
        B, V, _, _  = intrinsics.shape
        H, W = image_shape
        rendered_imgs, rendered_depths = [], []
        xyzs, opacitys, rotations, scales, features = gaussians.means, gaussians.opacities, gaussians.rotations, gaussians.scales, gaussians.harmonics.permute(0, 1, 3, 2).contiguous()
        covariances = gaussians.covariances
        for i in range(B):
            xyz_i = xyzs[i].float()
            feature_i = features[i].float()
            covar_i = covariances[i].float()
            scale_i = scales[i].float()
            rotation_i = rotations[i].float()
            opacity_i = opacitys[i].squeeze().float()
            test_w2c_i = extrinsics[i].inverse().float() # (V, 4, 4)
            test_intr_i_normalized = intrinsics[i].float()
            # Denormalize the intrinsics into standred format
            test_intr_i = test_intr_i_normalized.clone()
            test_intr_i[:, 0] = test_intr_i_normalized[:, 0] * W
            test_intr_i[:, 1] = test_intr_i_normalized[:, 1] * H
            sh_degree = (int(sqrt(feature_i.shape[-2])) - 1)
            # breakpoint()
            rendering, _, _ = rasterization(xyz_i, rotation_i, scale_i, opacity_i, feature_i,
                                                test_w2c_i, test_intr_i, W, H, sh_degree=sh_degree, 
                                                # near_plane=near[i].mean(), far_plane=far[i].mean(),
                                                render_mode="RGB+D", packed=False,
                                                near_plane=0.00001,
                                                backgrounds=self.background_color.unsqueeze(0).repeat(V, 1),
                                                radius_clip=0.1,
                                                covars=covar_i,
                                                rasterize_mode='classic') # (V, H, W, 3) 
            rendering_img, rendering_depth = torch.split(rendering, [3, 1], dim=-1)
            
            rendered_depths.append(rendering_depth.squeeze())
            rendered_imgs.append(rendering_img.permute(0, 3, 1, 2))
        return DecoderOutput(torch.stack(rendered_imgs), torch.stack(rendered_depths))
```

I also modify the code here: https://github.com/cvg/NoPoSplat/blob/9b04307ebe179d610c04208db8d69f7c3106d03b/src/model/types.py#L8

into the following to fit the type hint of the pytorch-lighting framework:

```
@dataclass
class Gaussians:
    means: Float[Tensor, "batch gaussian dim"]
    covariances: Float[Tensor, "batch gaussian dim dim"]
    harmonics: Float[Tensor, "batch gaussian 3 d_sh"]
    opacities: Float[Tensor, "batch gaussian"]
    scales: Float[Tensor, "batch gaussian 3"]
    rotations: Float[Tensor, "batch gaussian 4"]
```

And the code here: https://github.com/cvg/NoPoSplat/blob/9b04307ebe179d610c04208db8d69f7c3106d03b/src/model/encoder/encoder_noposplat.py#L223

into:

```
      return Gaussians(
          rearrange(
              gaussians.means,
              "b v r srf spp xyz -> b (v r srf spp) xyz",
          ),
          rearrange(
              gaussians.covariances,
              "b v r srf spp i j -> b (v r srf spp) i j",
          ),
          rearrange(
              gaussians.harmonics,
              "b v r srf spp c d_sh -> b (v r srf spp) c d_sh",
          ),
          rearrange(
              gaussians.opacities,
              "b v r srf spp -> b (v r srf spp)",
          ),
          rearrange(
              gaussians.scales,
              "b v r srf spp xyz -> b (v r srf spp) xyz",
          ),
          rearrange(
              gaussians.rotations,
              "b v r srf spp xyzw -> b (v r srf spp) xyzw",
          ),
      )
```


I believe here is the main modification of the code, if you need more information to help debugging, feel free to leave comment!
